Generate Java ops wrappers

karllessard opened the pull request:
This pull request generates Java wrappers for TensorFlow operations. Generated source files are compiled with the main Java source code into the `libtensorflow.jar` file. The process is done in two steps:

1- The op proto definition is parsed, with its api proto definitions if any, by the `op_parser`.
2- The parsed information is used to generated a Java wrapper class by the `op_generator`.

It is important to do this in two steps (instead of writing the source code as we parse) because any aspect found in the proto definition can affect the very beginning of the op class.

CC: @asimshankar

karllessard committed:
Parse op definition and generate a Java Op class.

karllessard committed:
Improve Javadoc and include first code review

asimshankar commented on tensorflow/java/src/gen/cc/op_parser.cc:
```diff
@@ -0,0 +1,417 @@
+  }
+  for (const auto& endpoint_def : op_api_.endpoint()) {
+    std::vector<string> name_tokens = str_util::Split(endpoint_def.name(), ".");
+    // if the endpoint specifies a package, use it, otherwise derive it from the
```
Instead of having `lib_name_` be based on the library name, how about we dump all into a single namespace (`org.tensorflow.ops`)? That may also simplify the BUILD rules as we won't need one per op library, we can just generate all operations in one go.

karllessard replied:
You mean all ops in a single Java package?

asimshankar commented on tensorflow/java/src/gen/cc/op_parser.h:
```diff
@@ -0,0 +1,137 @@
+  void add_attribute(const Operand& attribute) {
+    attributes_.push_back(attribute);
+  }
+  const std::vector<Operand>& options() const { return options_; }
```
How about calling these `required_attributes()` and `optional_attributes()` or `attributes()` and `optional_attributes()` - just to stick with the terminology used elsewhere.

karllessard replied:
The thing is that `options` is the idiomatic name in Java for `optional_attribute`. But I guess we can stick to `optional_attribute` then for the C++ part.

asimshankar reviewed (changes requested):
Thanks for the PR! As you suggested offline, I haven't looked at `op_generator.cc` yet but did look at the other files.

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -235,15 +216,42 @@ class Method {
+  }
+  const string& brief() const { return brief_; }
+  const string& details() const { return description_; }
+  Javadoc& details(const string description) {
```
oups...

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -235,15 +216,42 @@ class Method {
+    description_ = description;
+    return *this;
+  }
+  const std::list<std::pair<string, string>> tags() const { return tags_; }
```
oops again...

karllessard commented on tensorflow/java/src/gen/cc/op_gen_main.cc:
```diff
@@ -48,8 +48,11 @@ const char kUsageHeader[] =
-    "under which the generated subpackage and classes are to be located.\n\n";
+    "The '--base_package' overrides the default parent package under which "
+    "the generated subpackage and classes are to be located.\n\n"
+    "Finally, a list of directories of API proto definitions can be provided "
```
That's what I thought but looks like there is no group name in the `ApiDef` neither. I was hoping to find it in the `ApiDef_Endpoint.name()`, as the "Package" like it mentioned in the proto definition here:
>>   // Name should be either like "CamelCaseName" or
    // "Package.CamelCaseName". Client-language-specific ApiDefs may
    // use a snake_case convention instead of CamelCase.
    string name = 1;

but it looks that package is always empty and only the endpoint name part is present. Am I wrong?

asimshankar replied:
You're not wrong in that those aren't present yet.
But we should add them and use that mechanism to organize instead of lib names.
The Python API is organized like that with ApiDef overrides, for example: [1](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/python_api/api_def_DecodeJpeg.pbtxt), [2](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/python_api/api_def_BatchMatrixTriangularSolve.pbtxt), [3](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/python_api/api_def_RightShift.pbtxt) etc.

karllessard commented on tensorflow/java/src/gen/cc/op_gen_main.cc:
```diff
@@ -72,12 +75,15 @@ int main(int argc, char* argv[]) {
-  tensorflow::java::OpGenerator generator;
+  std::vector<tensorflow::string> api_dirs;
+  if (argc > 1) {
+    api_dirs = tensorflow::str_util::Split(argv[1], ",",
```
I'm good with this, to be honest I stole those `api_defs` arguments directly from the cc client but I agree it would be better to give them a name.

karllessard commented on tensorflow/java/src/gen/cc/op_parser.h:
```diff
@@ -0,0 +1,137 @@
+       description_(description), iterable_(iterable) {}
+    const string& graph_name() const { return graph_name_; }
+    const Variable& var() const { return var_; }
+    Variable* var_ptr() { return &var_; }
```
Oh yeah, I remember now... ;)

karllessard commented on tensorflow/java/src/gen/cc/op_parser.h:
```diff
@@ -0,0 +1,137 @@
+//
+// This object parses the definition and the api of an TensorFlow operation to
+// produce a specification that can be used for Java source code rendering.
+class OpParser {
```
Mmh, interesting... I only created a class because I did not want to carry the data shared by helper functions in all their signatures (`visited_attrs_`, `next_generic_`, etc). 

But I can check if that can be part of the `OpSpec` class directly instead.

asimshankar replied:
If you do need a class for that, I'd recommend hiding that as an implementation detail in an anonymous namespace in the `.cc` file and keep the header file more focused.

karllessard commented on tensorflow/java/src/gen/cc/op_parser.cc:
```diff
@@ -0,0 +1,417 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
haha

karllessard commented on tensorflow/java/src/gen/cc/op_parser.h:
```diff
@@ -0,0 +1,137 @@
+  void add_attribute(const Operand& attribute) {
+    attributes_.push_back(attribute);
+  }
+  const std::vector<Operand>& options() const { return options_; }
```
`options` are `optional_attributes`, I found that shorter names make the code cleaner overall. I'll add a comment.

karllessard commented on tensorflow/java/src/gen/cc/op_parser.cc:
```diff
@@ -0,0 +1,417 @@
+  return Type::Wildcard();
+}
+
+Type OpParser::DataTypeOf(const OpDef_AttrDef& attr, bool* iterable_out) {
```
I tried to avoir `TypeOf` because the type of an input, for example, can be `List<Operand<String>>` while the value return here is only the tensor's one (`String` in this case). Maybe `TensorTypeOf` then?

asimshankar replied:
I see. I think `TypeOf` with a comment is fine. Or maybe `TypeOfElement`?

karllessard commented on tensorflow/java/src/gen/cc/op_parser.cc:
```diff
@@ -0,0 +1,417 @@
+namespace java {
+namespace {
+
+string SnakeToCamelCase(const string& str, bool upper = false) {
```
I think that this method could be easily be part of TF `str_util.h`, the cc client already use a similar one. Wdyt?

karllessard commented:
Here @asimshankar , added your recommendations! BTW, I lost an incredible amount of time trying to implement a minimalist *markdown2javadoc* conversion utility, I think it would better go with a third party library, there is a few (e.g. `Discount`). What do you think?

karllessard committed:
Comments and little improvements to documentation parser

karllessard committed:
Improve again javadoc readability and quality

karllessard committed:
Simplify and improve generics handling in generator

asimshankar commented on tensorflow/java/src/gen/cc/source_writer.cc:
```diff
@@ -83,20 +83,20 @@ SourceWriter& SourceWriter::Append(const StringPiece& str) {
+    if (!type.parameters().empty()) {
+      Append("<");
+      for (const Type& t : type.parameters()) {
+        if (&t != &type.parameters().front()) {
```
The intention here is to avoid the first one, right? How about:

```c++
bool first = true;
for (const Type& t: type.parameters()) {
  if (!first) Append(", ");
  AppendType(t);
  first = false;
}
```

asimshankar commented on tensorflow/java/src/gen/cc/op_generator.cc:
```diff
@@ -14,53 +14,429 @@ limitations under the License.
+        const std::string api_def_file_pattern =
+            io::JoinPath(api_def_dir, "api_def_" + op.name() + ".pbtxt");
+        if (env_->FileExists(api_def_file_pattern).ok()) {
+          TF_CHECK_OK(api_map.LoadFile(env_, api_def_file_pattern));
```
Might want to add a `<< api_def_file_pattern` at the end so that if the check fails, we know which file it failed on.

asimshankar commented on tensorflow/java/src/gen/cc/op_generator.cc:
```diff
@@ -14,53 +14,429 @@ limitations under the License.
+    if (api_def->visibility() != ApiDef::SKIP) {
+      OpSpec op(OpSpec::Create(op_def, *api_def));
+      for (const EndpointSpec& endpoint : op.endpoints()) {
+        GenerateOp(op, endpoint, base_package, output_dir, env_, gmtime(&now));
```
I'd recommend against including the timestamp as an annotation in the generated files, that leads to spurious differences between identical builds.

asimshankar reviewed (changes requested):
Thanks Karl! `op_generator.cc` was nice and easy to follow. Regarding the markdown stuff, yeah, let's keep what you have for now and we can get fancier later.

karllessard commented on tensorflow/java/BUILD:
```diff
@@ -68,8 +68,13 @@ filegroup(
 )
 
+# Build the gen tool as a library, as it will be linked to a core/ops binary
+# files before making it an executable.
```
I think it is still safe to select explicitly the list of libraries that we want to link to when we generate the gen tool library, since some of them are more tricky to handle and have been excluded for now (e.g. `dataset_ops`).

And we still need the `java_op_gen_lib` to be compiled separately from the binary because it is linked with generator unit tests, like `source_writer_test`

But yes, I guess the `java_op_gen_tool` binary rule could be in BUILD instead and I'll do an overall review of the comments I left there.

asimshankar replied:
But we aren't linking specific libraries (and can handle exclusions with `ApiDef` files instead?) Anyway, seems like you're looking to clean this up so that's great. Thanks!

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Yeah, I wasn't sure about that neither, thank you ;)

karllessard commented on tensorflow/java/src/gen/cc/java_defs.h:
```diff
@@ -96,11 +102,10 @@ class Type {
-  Type& description(const string& description) {
-    description_ = description;
-    return *this;
+  const string full_name() const {
```
That was my first thought also but for some reason that I don't recall, I changed my mind... so I guess it wasn't important, I'll put back `canonical_name()` then!

karllessard commented on tensorflow/java/BUILD:
```diff
@@ -68,8 +68,13 @@ filegroup(
+    ],
     gen_base_package = "org.tensorflow.op",
     gen_tool = "java_op_gen_tool",
     ops_libs = [
```
See previous comment, we still need to exclude some lib. But that list of dependencies will be set explicitly in the `java_op_gen_tool` binary rule instead.

karllessard commented on tensorflow/java/src/gen/cc/op_specs.h:
```diff
@@ -0,0 +1,156 @@
+  //
+  // op_def: Op definition
+  // api_def: Op API definition
+  static OpSpec Create(const OpDef& op_def, const ApiDef& api_def);
```
There are a few reasons why it is better to do the ops generation in two phases (i.e. the parsing and the rendering). The main one is generics.

In order to know which attribute of an op is being used as a generic type for an argument, and which are gonna be inferred by inputs while others needs to be explicitly provided by the user, we need to go through all arguments/attributes of the ops before rendering the class.

Also, for each attributes identified to be generic, we need to keep track of its parameter name (e.g. the letter `U`) we assigned to it for other references, so there is a need to save parsed values in a stateful object like `*Spec`.

This is just an example, we also benefit from resolving only once arg/attr Java types and their Javadoc, since they are reused multiple times in the rendering phase.

asimshankar replied:
Hmm...I'm a little worried about the overall complexity here, but I admit I don't have a concrete suggestion on making this better. I guess keeping this `OpSpec` is fine for now.

asimshankar commented on tensorflow/java/src/gen/cc/op_specs.cc:
```diff
@@ -0,0 +1,413 @@
+  explicit TypeResolver(const OpDef& op_def) : op_def_(op_def) {}
+
+  Type TypeOf(const OpDef_ArgDef& arg_def, bool *iterable_out);
+  std::pair<Type, Type> TypeOf(const OpDef_AttrDef& attr_def,
```
A comment explaining the pair (what do the first and second element in the pair correspond to) would be nice.

Also, perhaps call the first one `TypeOf` and this one `TypesOf` since it returns two types.
(Otherwise it can be confusing that `TypeOf` returns either a single `Type` or an `std::pair` depending on the type of the arguments, which isn't obvious at the call sites).

asimshankar commented on tensorflow/java/src/gen/cc/op_specs.h:
```diff
@@ -0,0 +1,156 @@
+  const bool iterable_;
+};
+
+class AttributeSpec : public ArgumentSpec {
```
I'd recommend avoiding inheritance here and instead keeping the classes pure (non-virtual). Duplicating the 4 fields and getters should be fine here IMHO.

karllessard committed:
Second code review

karllessard commented:
Here @asimshankar , your last suggestions have been applied

asimshankar commented on tensorflow/core/api_def/java_api/api_def_FilterDataset.pbtxt:
```diff
@@ -0,0 +1,4 @@
+op {
```
Instead of this, would it make sense to have the `OpGenerator` just skip operations with `func` attributes for now? That might be simpler than white-listing? Later on, when function valued attributes are supported, we can revisit?

karllessard replied:
Well, looks like you saw right because a brand new op with a `func` attribute appeared when I last rebased, therefore was not skipped, and that broke the build. I agree to do a more robust solution for `func` attributes for now, I'll commit something.

asimshankar reviewed (dismissed):
Thanks!

karllessard committed:
Skip all ops with function attribute by default

karllessard commented:
Here @asimshankar , just added an explicit check to skip all operations with a `func` attribute.

I also removed the `java_api` folder for now, we'll populate all this at once in another PR, thanks!

karllessard committed:
Remove comment left-over

asimshankar reviewed (dismissed):
Thanks Karl!

karllessard commented:
@asimshankar , can you give me a hand here? Seems like now it is complaining that it can’t find my resource files but locally, I have no problem running it:
```
2018-05-05 17:22:45.430172: F tensorflow/java/src/gen/cc/source_writer.cc:70] Non-OK-status: ReadFileToString(env, fname, &data_) status: Not found: tensorflow/java/src/gen/resources/license.java.snippet; No such file or directory
```
Any clue how to fix this?

asimshankar commented:
Ah, that is because there is some missing data dependency in the BUILD rule (`bazel` aims for hermetic builds, so all the data and code dependencies need to be specified).

I would suggest that we do away with the data file dependency altogether and instead embed that copyright notice as a string constant in the generator. That way the `java_gen_op_tool` binary is a standalone binary and you don't have to worry about access to the data file every time it is invoked.

karllessard committed:
Embed Java license in generator code

karllessard commented:
Here @asimshankar , try #3. (I don't know if you have any trick to test locally what will be the output of the build server?)
